ray_on_golem/server/services/golem/golem.py

```diff
@@ -311,14 +319,77 @@ async def _upload_node_configuration(
         await context.run("mkdir -p /root/.ssh")
         await context.run(f'echo "{ssh_public_key_data}" >> /root/.ssh/authorized_keys')
 
-    async def _start_ssh_server(
-        self, context: WorkContext, ip: str, *, add_state_log: Callable[[str], Awaitable[None]]
-    ):
-        provider_desc = await self._get_provider_desc(context)
+    async def _start_ssh_server(self, context: WorkContext, ip: str):
+        provider_desc = await self.get_provider_desc(context.activity)
         logger.info("Starting ssh service on " f"{provider_desc}, {ip=}, {context.activity=}")
-        await add_state_log("[7/9] Starting ssh service...")
         await context.run("service ssh start")
 
+    async def _restart_ssh_server(self, context: WorkContext, ip: str):
+        provider_desc = await self.get_provider_desc(context.activity)
+        logger.debug(f"Restarting ssh service on {provider_desc}, {ip=}, {context.activity=}")
+        try:
+            await context.run("service ssh restart", timeout=120)
+            logger.debug(
+                f"Restarting ssh service on {provider_desc}, {ip=}, {context.activity=} done"
+            )
+        except Exception:
+            msg = f"Failed to restart SSH server {provider_desc}, {ip=}, {context.activity=}"
+            logger.warning(msg)
+            logger.debug(msg, exc_info=True)
+
+    @staticmethod
+    async def _verify_ssh_connection_check(activity_id, ssh_command):
```

Comment:
this function doesn't do what it says it's doing :D

maybe we should move this:
```
        ssh_command = (
            f"{get_ssh_command(ip, ssh_proxy_command, ssh_user, ssh_private_key_path)} uptime"
        )
```

inside this function and only pass the actual command there... otherwise, it's just a generic process runner which checks the process' exit code ;)

Reply:
fixed